src/main/java/org/opensearch/ad/ratelimit/ConcurrentWorker.java

```diff
@@ -0,0 +1,167 @@
+     * @param maintenanceFreqConstant a constant help define the frequency of maintenance.  We cannot do
+     *   the expensive maintenance too often.
+     * @param clientUtil utility with ES client
+     * @param concurrencySetting Max concurrent processing of the queued events
+     * @param executionTtl Max execution time of a single request
+     * @param stateTtl max idle state duration.  Used to clean unused states.
+     * @param nodeStateManager node state accessor
+     */
+    public ConcurrentWorker(
+        String queueName,
+        long heapSizeInBytes,
+        int singleRequestSizeInBytes,
+        Setting<Float> maxHeapPercentForQueueSetting,
+        ClusterService clusterService,
+        Random random,
+        ADCircuitBreakerService adCircuitBreakerService,
+        ThreadPool threadPool,
+        Settings settings,
+        float maxQueuedTaskRatio,
+        Clock clock,
+        float mediumSegmentPruneRatio,
+        float lowSegmentPruneRatio,
+        int maintenanceFreqConstant,
+        ClientUtil clientUtil,
+        Setting<Integer> concurrencySetting,
+        Duration executionTtl,
+        Duration stateTtl,
+        NodeStateManager nodeStateManager
+    ) {
+        super(
+            queueName,
+            heapSizeInBytes,
+            singleRequestSizeInBytes,
+            maxHeapPercentForQueueSetting,
+            clusterService,
+            random,
+            adCircuitBreakerService,
+            threadPool,
+            settings,
+            maxQueuedTaskRatio,
+            clock,
+            mediumSegmentPruneRatio,
+            lowSegmentPruneRatio,
+            maintenanceFreqConstant,
+            stateTtl,
+            nodeStateManager
+        );
+
+        this.permits = new Semaphore(concurrencySetting.get(settings));
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(concurrencySetting, it -> permits = new Semaphore(it));
+        this.clientUtil = clientUtil;
+
+        this.lastExecuteTime = clock.instant();
+        this.executionTtl = executionTtl;
+    }
+
+    @Override
+    public void maintenance() {
+        super.maintenance();
+
+        if (lastExecuteTime.plus(executionTtl).isBefore(clock.instant()) && permits.availablePermits() == 0 && false == isQueueEmpty()) {
+            LOG.warn("previous execution has been running for too long.  Maybe there are bugs.");
+
+            // Release one permit.  This is a stop gap solution as I don't know
+            // whether the system is under heavy workload or not. Release multiple
+            // permits might cause the situation even worse. So I am conservative here.
+            permits.release();
+        }
+    }
+
+    /**
+     * try to execute queued requests if there are concurrency slots and return right away.
+     */
+    @Override
+    protected void triggerProcess() {
+        threadPool.executor(AnomalyDetectorPlugin.AD_THREAD_POOL_NAME).execute(() -> {
+            if (permits.tryAcquire()) {
```

Comment:
Just a question, the number of semaphores should be less than or equal to the number of threads, right?

Reply:
seems not. I have seen some well-known project using 10k semaphores to represent total number of tokens that account for total heap memory available. Each mb is 1 semaphore.

I can have 100 semaphores with 1 thread.  If all of the code are asynchronous, those 100 operations (protected by the semaphores) can still run concurrently as long as they don't use the thread at the same time.